### ISSUE TITLE
NP lookup always performed regardless of state of override_npdi option

### ISSUE THREAD

**Issue by dcollins991** [Metaswitch/sprout#1096]
For the Number Portability Feature Spec The configuration option, override_npdi, is added. If this is set to true, then the I-CSCF, S-CSCF and BGCF will perform an NP lookup and processing even when the request URI has npdi set. This defaults to false.

Regardless of whether the parameter is set to true or false an NP lookup always occurs.  I tried the following forms for telephone numbers.

To: <sip:3135551002@vm6.richlab.datcon.co.uk;user=phone;pndi>
To: <tel:+2012040002;pndi>
To: <tel:2012040002;phone-context=192.168.165.150;pndi>

**Comment by dcollins991** [comment]
Sorry, should be ;npdi above.

**Comment by mirw** [comment]
@dcollins991, the `npdi` parameter should be applied to the request URI, not the To header.  Please can you check that and, if it still occurs, gather SAS trace?  Thanks!

**Comment by dcollins991** [comment]
It appears in both.

**Comment by dcollins991** [comment]
Looking at RFC 4694 it appears wherever the tel uri appears.  Right?

**Comment by mirw** [comment]
Yes, it can appear wherever the TEL URI appears, but we'll only look at the value in the request URI - the To header is off-limits to the S-CSCF.

One other thing to note - the SIP example you have above is wrong - the npdi parameter should be attached to the user part, i.e. `sip:3135551002;npdi@vm6.richlab.datcon.co.uk;user=phone`.

**Comment by dcollins991** [comment]
OK.  Will recheck the uri generation in the test case.

**Comment by dcollins991** [comment]
This scenario has the INVITE as,
INVITE sip:3135551002;npdi@vm6.richlab.datcon.co.uk;user=phone SIP/2.0

The npdi_override is set to false.

Yet, the ENUM translation dip is still performed.

Now there appears to be a second problem in the parseing of the domain within the NAPTR record.  The NAPTR record is,
2.0.0.1.5.5.5.3.1.3.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!sip:4175191002;npdi@vm6.richlab.datcon.co.uk!" .

Now my expectation here is if you do perform the ENUM lookup you would just replace the tel uri with the sip uri and then lookup the domain, vm6.richlab.datcon.co.uk.  But it has an error as,


015-06-16 12:39:41.966sprout@sprout-1.vm6.richlab.datcon.co.ukENUM (DNS NAPTR) request sent for domain 2.0.0.1.5.5.5.3.1.3.e164.arpa
2015-06-16 12:39:41.967sprout@sprout-1.vm6.richlab.datcon.co.ukENUM (DNS NAPTR) response received for domain 2.0.0.1.5.5.5.3.1.3.e164.arpa
2015-06-16 12:39:41.967sprout@sprout-1.vm6.richlab.datcon.co.ukENUM translated 313-555-1002 to sip:4175191002;npdi@vm6.richlab.datcon.co.uk
2015-06-16 12:39:41.968sprout@sprout-1.vm6.richlab.datcon.co.ukBGCF lookup for domain "" failed - routing directly 

The bgcf.json file has,
{
    "routes" : [
        {   "name" : "Route to vcfs",
            "domain" : "cfs.vm6.richlab.datcon.co.uk",
            "route" : ["sip:192.168.165.109:5070;transport=TCP"]
        },
        {   "name" : "Route to vm6", 
            "domain" : "vm6.richlab.datcon.co.uk",
            "route" : ["sip:192.168.165.109:6060;transport=TCP"]
        },
        {   "name" : "Route to mgcf",
            "domain" : "192.168.165.150",
            "route" : ["sip:192.168.165.150:5060"]
        },
        {   "name" : "Route to vcfs for np",
            "number" : "2012040002",
            "route" : ["sip:192.168.165.109:5070;transport=TCP"]
        }
    ]
}

**Comment by dcollins991** [comment]
I can get the call to work if I set a domain name equal to "" in the bgcf.json file.  However, the npdi_override flag does not have any effect.

**Comment by rkd-msw** [comment]
I think this might be working as expected:
* Sprout has a URI of sip:3135551002;npdi@vm6.richlab.datcon.co.uk;user=phone
* It does the ENUM lookup and gets back sip:4175191002;npdi@vm6.richlab.datcon.co.uk
* Because npdi was set in the original URI, that isn't overriden, and we keep using 'sip:3135551002;npdi@vm6.richlab.datcon.co.uk;user=phone'
* When we come to do BGCF routing on that, because 'user=phone' is set, the domain is ignored - that's why it only matches a domain of "" or "*"

Does that describe what you're seeing? Could you upload a SAS trace?

**Comment by rkd-msw** [comment]
(And I think if npdi_override was true, you'd ignore the npdi flag, use the ENUM-provided `sip:4175191002;npdi@vm6.richlab.datcon.co.uk` URI, and route there.)

**Comment by rkd-msw** [comment]
Having discussed with @mirw, we should have better documentation of this.

**Comment by dcollins991** [comment]
See the SAS trace in issue1096/empty_domain3135551002.sas.

**Comment by rkd-msw** [comment]
Yes, it looks from that SAS trace as though we are honoring the npdi flag and not using the value returned from ENUM - the subsequent INVITE is still using the original URI, `INVITE sip:3135551002;npdi@vm6.richlab.datcon.co.uk;user=phone SIP/2.0`.

To be clear, having `npdi` in a URI doesn't stop an ENUM lookup from being done - it just means that, if the ENUM lookup just returns number portability information, we'll ignore it. (But we would stiull use the result if it was a real SIP URI and not number portability information.)

Are you happy that this is working as expected, and that this issue should just cover documentation updates and clearer SAS logs?

**Comment by dcollins991** [comment]
Let me test it and will respond shortly.

**Comment by dcollins991** [comment]
I thought the purpose of the npdi flag was to save time and not do an enum dip.

**Comment by rkd-msw** [comment]
No, I think TS 24.229 is pretty clear on this:

```
If the above translation from an international public telecommunications number to a SIP URI failed, but NP data was obtained from the database and there is no "npdi" parameter in the received request, then the S-CSCF shall, based on operator policy, replace the URI in the Request-URI with the obtained NP data, prior to forwarding the request to the BGCF or other appropriate entity. If the received request already contains a tel-URI "npdi" parameter, then the S-CSCF may update the URI with the obtained NP data.
```

So you always try and translate it to a SIP URI. If you don't get back a normal SIP URI, but get back a URI with NP data (`rn` or `npdi` parameters), then you decide whether to use that data based on the presence of `npdi`.

Is that OK?

**Comment by dcollins991** [comment]
For, INVITE tel:+2012040002;npdi SIP/2.0 I still get an empty domain.
Similarly for,
INVITE tel:2012040002;phone-context=192.168.165.150;npdi SIP/2.0
INVITE tel:+12012040002;npdi SIP/2.0

All return an empty domain.  

I can provide SAS traces.

The NAPTR records are:
2.0.0.0.4.0.2.1.0.2.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!sip:2012040002;npdi;rn=2012040002@cfs.vm6.richlab.datcon.co.uk!" .
2.0.0.0.4.0.2.1.0.2.1.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!sip:2012040002;npdi;rn=2012040002@cfs.vm6.richlab.datcon.co.uk!" .
1.0.0.1.5.5.5.3.1.3.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!tel:+1-313-555-0101;npdi;rn=4175191001!" .
2.0.0.1.5.5.5.3.1.3.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!sip:4175191002;npdi@vm6.richlab.datcon.co.uk!" .
3.0.0.1.5.5.5.3.1.3.1.e164.arpa. NAPTR 10 100 "u" "E2U+pstn:tel" "!^(.*)$!sip:4175191002;npdi@vm6.richlab.datcon.co.uk!" .
*.6.5.5.3.1.3.e164.arpa. NAPTR 10 100 "u" "E2U+sip" "!^(.*)$!sip:\\1@vm6.richlab.datcon.co.uk!" .

**Comment by dcollins991** [comment]
OK to the standard quote above.

**Comment by rkd-msw** [comment]
Yes, that's what I'd expect:
* All those ENUM lookups contain routing information, so the npdi flags mean they won't be used
* We're therefore routing based on the original TEL URI - because it's a TEL URI it naturally has no domain (which shows up as an empty domain in the logs)

**Comment by dcollins991** [comment]
I have override_npdi set to true  for the above cases.  Can I call you?

**Comment by rkd-msw** [comment]
With `override_npdi` set, those URIs should get rewritten to `sip:2012040002;npdi;rn=2012040002@cfs.vm6.richlab.datcon.co.uk`, and the BGCF should route on the `rn` parameter - you should see a SAS log with the `rn` (2012040002) and not an empty domain.

Can you provide:
* SAS trace for those two calls
* a diags bundle so I can check override_npdi is being set appropriately?

We've agreed that this isn't high-priority.

**Comment by dcollins991** [comment]
I have gathered the diags for sprout-1 and sprout 2.  There are 3 SAS files: BGCFLookupNoNPDIinReqURI.sas.BGCFLookupNPDIinReqURI.sas, and another one with a tel uri that fails as expected.  The first two have user=phone.  Let me know if there are any issues or you need any other traces.

**Comment by bossmc** [comment]
Retesting is blocked on #1120 and #1121.  Passing to @rkd-msw to confirm when testing is unblocked.

**Comment by bossmc** [comment]
Raising priority as this might be a spec-compliance/functional behaviour issue.

**Comment by rkd-msw** [comment]
@dcollins991 - in release-82, I've reworked this behaviour, and enhanced the SAS logs. The new SAS logs (https://github.com/Metaswitch/clearwater-sas-resources/commit/9ce5c167bbb721de7f0f2037e89b0be4ad85bc9d#diff-d27b7cf4f49cf128e646ccab9a8dd245) should be much more helpful in understanding what's going on and why.

Can you retest on release-82, and pass this issue back to me with SAS traces for any call flows which you still think don't make sense?

**Comment by sathiyan-sivathas** [comment]
After speaking to @rkd-msw, we've agreed that this has been fixed by #1120 and #1121 and that we should close this.
